filters/include/pcl/filters/functor_filter.h

```diff
@@ -76,22 +77,40 @@ class FunctorFilter : public FilterIndices<PointT> {
    */
   void
   applyFilter(Indices& indices) override
+  {
+    applyFilter(executor::best_fit{}, indices);
+  }
+
+  template <typename Executor, typename std::enable_if_t<executor::is_instance_of_base_v<executor::inline_executor, Executor> || executor::is_instance_of_base_v<executor::omp_executor, Executor>, int> = 0>
+  void
+  applyFilter (Executor &&exec, Indices& indices)
   {
     indices.clear();
-    indices.reserve(input_->size());
+    indices.reserve(indices_->size());
+    std::vector<int> keep(indices_->size(), 0);
+    std::vector<int> remove(indices_->size(), 0);
+
     if (extract_removed_indices_) {
       removed_indices_->clear();
-      removed_indices_->reserve(input_->size());
+      removed_indices_->reserve(indices_->size());
     }
 
-    for (const auto index : *indices_) {
-      // functor returns true for points that should be selected
-      if (negative_ != functor_(*input_, index)) {
-        indices.push_back(index);
-      }
-      else if (extract_removed_indices_) {
-        removed_indices_->push_back(index);
-      }
+    auto cond_filtering = [&](index_t idx) {
+      exec.execute([&]() {
+        if (negative_ != functor_(*input_, idx)) {
+          keep[idx] = true;
+        }
+        else if (extract_removed_indices_) {
+          remove[idx] = true;
+        }
+      });
+    };
+
+    exec.bulk_execute(cond_filtering, indices_->size());
+
+    for (index_t i = 0; i < keep.size(); ++i) {
```

Comment:
I'd advise to split this merge part into another function. Possible speed-ups:
* with SSE: mask based copy
* with OMP and CUDA: `O(log2(keep.size())` reduce copy
* with OMP and CUDA: independent indices and removed_indices

Reply:
I didn't get the last point

Reply:
You can have two loops, one for appending to `indices` and one for appending to `removed_indices_`. Both are independent

Reply:
Wouldn't that cause a slowdown, since it now has two iterate twice? While currently, only and additional check is added to the loop, which should be taken care of branch prediction.